pkg/controller/apiextensions/offered/watch.go

```diff
@@ -84,3 +90,61 @@ func addClaim(obj runtime.Object, queue adder) {
 		queue.Add(reconcile.Request{NamespacedName: meta.NamespacedNameOf(cp.GetClaimReference())})
 	}
 }
+
+// EnqueueRequestForControllersClaim enqueues a reconcile.Request for the
+// NamespacedName of a Secret's controller's ClaimReference. The controller must
+// be a composite resource.
+type EnqueueRequestForControllersClaim struct{ client client.Reader }
+
+// Create adds a NamespacedName for the supplied CreateEvent if its Object is
+// the connection secret of a claimed composite resource.
+func (e *EnqueueRequestForControllersClaim) Create(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+	e.addClaim(evt.Object, q)
+}
+
+// Update adds a NamespacedName for the supplied UpdateEvent if its Objects are
+// the connection secret of a claimed composite resource.
+func (e *EnqueueRequestForControllersClaim) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	e.addClaim(evt.ObjectOld, q)
+	e.addClaim(evt.ObjectNew, q)
+}
+
+// Delete adds a NamespacedName for the supplied DeleteEvent if its Object is
+// the connection secret of a claimed composite resource.
+func (e *EnqueueRequestForControllersClaim) Delete(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	e.addClaim(evt.Object, q)
+}
+
+// Generic adds a NamespacedName for the supplied GenericEvent if its Object is
+// the connection secret of a claimed composite resource.
+func (e *EnqueueRequestForControllersClaim) Generic(evt event.GenericEvent, q workqueue.RateLimitingInterface) {
+	e.addClaim(evt.Object, q)
+}
+
+func (e *EnqueueRequestForControllersClaim) addClaim(obj runtime.Object, queue adder) {
+	s, ok := obj.(*corev1.Secret)
+	if !ok {
+		return
+	}
+
+	if s.Type != resource.SecretTypeConnection {
+		return
+	}
+
+	ref := v1.GetControllerOf(s)
+	if ref == nil {
+		return
+	}
+
+	nn := types.NamespacedName{Name: ref.Name}
+	cp := &composite.Unstructured{}
+	cp.SetAPIVersion(ref.APIVersion)
+	cp.SetKind(ref.Kind)
+	if err := e.client.Get(context.TODO(), nn, cp); err != nil {
+		return
+	}
```

Comment:
I don't love that we need a client here to get the secret's XR in order to resolve its claim ref - especially since we have no context and no opportunity to return an error here -  but at least we should be reading from cache. 

I keep going back and forth on whether we should instead annotate XR connection secrets with their claim's GVK, namespace, and name. This would allow us to skip the intermediate get of the XR and just enqueue a request per the secret's annotations. Performance wise they're about the same - the main reason to go with this approach is that it lets us make the connection with the information we already have rather than adding new information (i.e. annotations).

CC @muvaf what's your opinion here?

Reply:
Yeah, I think that sounds like a good compromise.

Ideally, I'd want XRC controller to be notified when XR changes, which would be caused by a change in secret so that the information chain is still straight and XR secret doesn't need to know about XRC. But there doesn't seem to be a sane way to do achieve that.

Reply:
> Ideally, I'd want XRC controller to be notified when XR changes, which would be caused by a change in secret so that the information chain is still straight and XR secret doesn't need to know about XRC. But there doesn't seem to be a sane way to do achieve that.

One approach here could be to modify the XR somehow when its connection secret changes. i.e.:

1. XRC controller is watching XR.
1. XR controller is watching XR connection secret.
1. When XR connection secret changes, we set `status.connectionDetails.lastRefreshTime` or something.

This way the change to the secret would trigger a change to the XR which would trigger a change to the XRC.

Reply:
> When XR connection secret changes

That's kind of hard to detect though unless we're willing to assume _every_ event is a change, which could potentially render too frequent reconciliations.

FWIW, we'll call `GET` for every connection secret with an owner, which is most of what we publish, so it will be called very frequently in different go routines. Hope the cache helps for not hitting the api-server for those calls. It could really hog it after a few composite resources.

Reply:
> That's kind of hard to detect though unless we're willing to assume every event is a change, which could potentially render too frequent reconciliations.

Yeah. I'm thinking we'd have the XR reconciler report this, i.e. update the connection secret publishing behaviour to report whether the secret actually changed. That should be doable, but it's definitely a bigger diff to do it. I'm still on the fence so I might take a shot at that implementation in a competing PR so we can compare.

Reply:
Okay, I've opened https://github.com/crossplane/crossplane/pull/1927 which uses the implementation I described. In the end I much prefer that implementation (despite it being a larger diff):

* It feels more elegant than having to make an API call (even to cache) in order to map secrets to XRs.
* Including the time we last published connection details in the status seems handy for debugging.
* Watching connection secret changes indirectly via their associated XR is more future proof, as it would allow us to 'watch' for connection detail changes even if those details were stored in an external system such as Vault, rather than a Secret.

Reply:
Yeah, I like that implementation more, too. It seems you overcame the frequent event problem by comparing the content of the secret. So, we'll get an event only if the content of the secret changes, which is precisely what we care.